keps/sig-scheduling/4671-gang-scheduling/README.md

```diff
@@ -0,0 +1,983 @@
+In further phases, we may decide to relax validation and allow for setting some of the fields later.
+Moreover, the visibility into issues (debuggability) will depend on [#5510], but we don't
+treat it as a blocker.
+
+[#5510]: https://github.com/kubernetes/enhancements/pull/5510
+
+The example below shows how this could look like for with the following `Workload` object:
+
+```yaml
+apiVersion: scheduling/v1alpha1   
+kind: Workload
+metadata:
+  name: jobset
+spec:
+  podGroups:
+    - name: "job-1"
+      replicas: 4
+      policy:
+        gang:
+          minCount: 100
+```
+
+```yaml
+apiVersion: v1
+kind: Pod
+name:
+  jobset-job-1-abc123
+spec:
+  ...
+  workload:
+    name: jobset
+    podGroup: job-1
+    podGroupReplicaIndex: 2
+  ...
+
+```
+
+We decided for this option because it is more succint and makes the role of a pod clear just
+from inspecting the pod (and simple/efficient to group).
+We acknowledge the fact that this option may require additional minor changes in the controllers
+to adopt this pattern (e.g. for LeaderWorkerSet we will need to populate the pod template
+similarly that we currently populate the labels).
+
+The primary alternative we consider was to introduce the the `PodGroupSelector` on each `PodGroup`
+to identify pods belonging to it. However, with this pattern:
+- there are additional corner cases (e.g. a pod links to a workload but none of its PodGroups matching
+  that pod)
+- for replicated gang, we can't use the full label selector, but rather support specifying only the
+  label key, similar to `MatchLabelKeys` in pod affinity
+
+
+### API
+
+The `Workload` type will be defined with the following structure:
+```go
+type Workload struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+	Spec WorkloadSpec
+	Status WorkloadStatus
+}
+
+// WorkloadSpec describes a workload in a portable way that scheduler and related
+// tools can understand.  
+type WorkloadSpec struct {
+    // ControllerRef points to the true workload, e.g. Deployment.
+    // It is optional to set and is intended to make this mapping easier for
+    // things like CLI tools.
+    // This field is immutable.
+    ControllerRef *v1.ObjectReference
+
+    // PodGroups is a list of groups of pods.
+    // Each group may request gang scheduling.
+    PodGroups []PodGroup 
+}
+
+// PodGroup is a group of pods that may contain multiple shapes (PodSets) and may contain
+// multiple dense indexes (PodSubGroups) and which can optionally be replicated in a variable
+// number of identical copies.
+type PodGroup struct {
+    Name *string
+
+    // Number of identical instances of PodGroup that are part of the Workload.
+    // Defaults to 1.
+    Replicas int
+
+    // Policy defines the configuration of the PodGroup to enable different
+    // scheduling policies.
+    Policy PodGroupPolicy
+}
+
+// PodGroupPolicy defines scheduling configuration of a PodGroup.
+type PodGroupPolicy struct {
+    // Exactly one of the policies should be set.
```

Comment:
Nit: from api perspective you'll want to add a discriminator field, expressing which policy is at play. That's what we've been doing with all the union-like types.